Fixes #834 Major flavor testing revamp

MichaelStritt opened the pull request:
### Linked issue

Read description in #834.

### How to test

Check testing description in #834.

### Comments

Let's improve this step by step.

MichaelStritt commented:
### Up-to-date exemplary output of bugs 🐛 

```
>> flavors.loggingTable
ans =
  4×3 table

           message                             stack                                                name                                                      
    ______________________    _______________________________________    ____________________________________________________

    'this is a test error'    'xASL_module_Structural: line 42, ...'    '...\FlavorDatabase\GE_PCASL_3Dspiral_14.0LX_1'    
    'this is a test error'    'xASL_module_ASL: line 36, ...'           '...\FlavorDatabase\GE_PCASL_3Dspiral_14.0LX_1'    
    'this is a test error'    'xASL_module_Structural: line 42, ...'    '...\FlavorDatabase\GE_PCASL_3Dspiral_15.0LX_WIP_1'
    'this is a test error'    'xASL_module_ASL: line 36, ...'           '...\FlavorDatabase\GE_PCASL_3Dspiral_15.0LX_WIP_1'
```

jan-petr reviewed (changes requested):
I fixed two small bugs. All references are now correctly pushed. It all runs through without issues.

But:
1. When comparing Legacy with reference. Can we skip log files? Cause I didn't upload logs to reference. `Missing:   /ExploreASL/xASL_module_Import.log`
2. At the end - the log returned `no errors or warnings` - which is strange or have we removed all the errors and warnings already?!

Other than that - ready to merge this one in the current state and do more changes elsewhere. Because we need to be able to work with new Flavors and Testing to fix other issues.

MichaelStritt commented:
@jan-petr 

>     1. When comparing Legacy with reference. Can we skip log files? Cause I didn't upload logs to reference. `Missing:   /ExploreASL/xASL_module_Import.log`

Yeah, there are two solutions. Easiest is to create an optional input to ignore all .log files. The other method would be to return all differences in a struct and remove the log file differences there. Wasn't sure which one would be optimal. I definitely agree that this is necessary though. I removed the import logs as well when I was uploading the references.

>     2. At the end - the log returned `no errors or warnings` - which is strange or have we removed all the errors and warnings already?!

I think the BIDS comparison does not work 100% correctly. Don't know if we should work on this here though. I saw for one text file that they were not the same, it did report it, but the "identical" value was still "identical". I want to integrate this in a nicer way anyway, so maybe we do this later on. I just thought we need an initial new running script for all of this. Maybe we also delete `xASL_test_IntegrationTesting` while we're at it, because that's where some of those methods came from. There I was doing all of this a little bit differently, but I came up with this loggingTable stuff which can be quite useful. Right now we definitely don't return errors in the import, so at least the `flavors.loggingTable` wont return those buggy TopUp datasets and so on.

> Other than that - ready to merge this one in the current state and do more changes elsewhere. Because we need to be able to work with new Flavors and Testing to fix other issues.

Agree, that's why this was the first thing I wanted to work on 👍

MichaelStritt commented:
@jan-petr 

## Example with subset of flavors

### Exemplary JSON file

```json
{
   "pathExploreASL":     "...//ExploreASL",
   "pathFlavorDatabase": "...//FlavorDatabase",
   "cmdCloneFlavors":    "git clone git@github.com:ExploreASL/FlavorDatabase.git",
   "flavorList":         ["GE_PCASL_3Dspiral_14.0LX_1", "Philips_PCASL_2DEPI_3.2.1.1_1", "Siemens_PASL_3DGRASE_E11_1"]
}
```

### Test errors in import and processing

```
flavors.loggingTable
ans =
  12×3 table

            message                                                                                                                                                  stack                                                                                                                                                       name              
    _______________________    _________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________    _______________________________

    'test error import'        'xASL_module_Import: line 226, ExploreASL_ImportMaster: line 36, ExploreASL_Master: line 71, ExploreASL: line 14, xASL_test_Flavors_DCM2BIDS: line 47, xASL_test_Flavors: line 74, xASL_test_FullPipelineTest: line 107'                                                             'GE_PCASL_3Dspiral_14.0LX_1'   
    'test error import'        'xASL_module_Import: line 226, ExploreASL_ImportMaster: line 36, ExploreASL_Master: line 71, ExploreASL: line 14, xASL_test_Flavors_DCM2BIDS: line 59, xASL_test_Flavors: line 74, xASL_test_FullPipelineTest: line 107'                                                             'GE_PCASL_3Dspiral_14.0LX_1'   
    'test error import'        'xASL_module_Import: line 226, ExploreASL_ImportMaster: line 36, ExploreASL_Master: line 71, ExploreASL: line 14, xASL_test_Flavors_DCM2BIDS: line 47, xASL_test_Flavors: line 74, xASL_test_FullPipelineTest: line 107'                                                             'Philips_PCASL_2DEPI_3.2.1.1_1'
    'test error import'        'xASL_module_Import: line 226, ExploreASL_ImportMaster: line 36, ExploreASL_Master: line 71, ExploreASL: line 14, xASL_test_Flavors_DCM2BIDS: line 59, xASL_test_Flavors: line 74, xASL_test_FullPipelineTest: line 107'                                                             'Philips_PCASL_2DEPI_3.2.1.1_1'
    'test error import'        'xASL_module_Import: line 226, ExploreASL_ImportMaster: line 36, ExploreASL_Master: line 71, ExploreASL: line 14, xASL_test_Flavors_DCM2BIDS: line 47, xASL_test_Flavors: line 74, xASL_test_FullPipelineTest: line 107'                                                             'Siemens_PASL_3DGRASE_E11_1'   
    'test error import'        'xASL_module_Import: line 226, ExploreASL_ImportMaster: line 36, ExploreASL_Master: line 71, ExploreASL: line 14, xASL_test_Flavors_DCM2BIDS: line 59, xASL_test_Flavors: line 74, xASL_test_FullPipelineTest: line 107'                                                             'Siemens_PASL_3DGRASE_E11_1'   
    'test error structural'    'xASL_module_Structural: line 42, runIteration: line 394, xASL_Iteration: line 100, ExploreASL_ProcessMaster: line 77, ExploreASL_Master: line 85, ExploreASL: line 14, xASL_test_Flavors_ExploreASL: line 13, xASL_test_Flavors: line 100, xASL_test_FullPipelineTest: line 119'    'GE_PCASL_3Dspiral_14.0LX_1'   
    'test error asl'           'xASL_module_ASL: line 36, runIteration: line 394, xASL_Iteration: line 100, ExploreASL_ProcessMaster: line 103, ExploreASL_Master: line 85, ExploreASL: line 14, xASL_test_Flavors_ExploreASL: line 13, xASL_test_Flavors: line 100, xASL_test_FullPipelineTest: line 119'          'GE_PCASL_3Dspiral_14.0LX_1'   
    'test error structural'    'xASL_module_Structural: line 42, runIteration: line 394, xASL_Iteration: line 100, ExploreASL_ProcessMaster: line 77, ExploreASL_Master: line 85, ExploreASL: line 14, xASL_test_Flavors_ExploreASL: line 13, xASL_test_Flavors: line 100, xASL_test_FullPipelineTest: line 119'    'Philips_PCASL_2DEPI_3.2.1.1_1'
    'test error asl'           'xASL_module_ASL: line 36, runIteration: line 394, xASL_Iteration: line 100, ExploreASL_ProcessMaster: line 103, ExploreASL_Master: line 85, ExploreASL: line 14, xASL_test_Flavors_ExploreASL: line 13, xASL_test_Flavors: line 100, xASL_test_FullPipelineTest: line 119'          'Philips_PCASL_2DEPI_3.2.1.1_1'
    'test error structural'    'xASL_module_Structural: line 42, runIteration: line 394, xASL_Iteration: line 100, ExploreASL_ProcessMaster: line 77, ExploreASL_Master: line 85, ExploreASL: line 14, xASL_test_Flavors_ExploreASL: line 13, xASL_test_Flavors: line 100, xASL_test_FullPipelineTest: line 119'    'Siemens_PASL_3DGRASE_E11_1'   
    'test error asl'           'xASL_module_ASL: line 36, runIteration: line 394, xASL_Iteration: line 100, ExploreASL_ProcessMaster: line 103, ExploreASL_Master: line 85, ExploreASL: line 14, xASL_test_Flavors_ExploreASL: line 13, xASL_test_Flavors: line 100, xASL_test_FullPipelineTest: line 119'          'Siemens_PASL_3DGRASE_E11_1'   
```

jan-petr commented on Functions/xASL_bids_CompareStructuresJSON.m:
```diff
@@ -0,0 +1,222 @@
+% FORMAT: [differences,identical,dn] = xASL_bids_CompareStructuresJSON(differences,identical,bPrintReport,allFiles,iFile,dn,currentFileA,currentFileB)
+%
+% INPUT:
+%         ...
```
sorry

MichaelStritt replied:
Easy fix 04b6f5a 👍

jan-petr commented on Functions/xASL_bids_CompareStructuresNIFTI.m:
```diff
@@ -0,0 +1,55 @@
+% FORMAT: [differences,identical,dn] = xASL_bids_CompareStructuresNIFTI(differences,identical,bPrintReport,detailedOutput,allFiles,iFile,dn,currentFileA,currentFileB,threshRmseNii)
+%
+% INPUT:
+%         ...
```
sorry

MichaelStritt replied:
No worries 😉 71a97ec

jan-petr commented on Functions/xASL_bids_CompareStructuresTEXT.m:
```diff
@@ -0,0 +1,39 @@
+% FORMAT: [differences,identical,dn] = xASL_bids_CompareStructuresTEXT(differences,identical,bPrintReport,allFiles,iFile,dn,currentFileA,currentFileB)
+%
+% INPUT:
+%         ...
```
sorry

MichaelStritt replied:
Easy fix 04b6f5a 👍

jan-petr commented on Functions/xASL_bids_CompareStructuresTSV.m:
```diff
@@ -0,0 +1,51 @@
+% FORMAT: [differences,identical,dn] = xASL_bids_CompareStructuresTSV(differences,identical,bPrintReport,allFiles,iFile,dn,currentFileA,currentFileB)
+%
+% INPUT:
+%         ...
```
becoming boring :)

MichaelStritt replied:
Easy fix 04b6f5a 👍

jan-petr commented on Functions/xASL_test_Flavors.m:
```diff
@@ -1,26 +1,27 @@
 %xASL_test_Flavors Runs the complete testing of Flavors including import from DICOM to BIDS, processing and comparison
 %
-% FORMAT: xASL_test_Flavors(pathExploreASL, pathTest[, bTest, x])
+% FORMAT: xASL_test_Flavors(pathExploreASL, pathFlavorDatabase[, bTest, x])
```
sorry

MichaelStritt replied:
Easy fix 04b6f5a 👍

jan-petr commented on Functions/xASL_test_Flavors_DCM2BIDS.m:
```diff
@@ -1,12 +1,17 @@
 %
 % INPUT:
-%   baseDirImport    - Directory for import - sourcedata files are in the 'sourcedata' folder (REQUIRED)
+%   testConfig         - Struct which contains the paths to the ExploreASL
```
FORMAT above

MichaelStritt replied:
Easy fix 04b6f5a 👍

jan-petr reviewed (changes requested):
Minor header stuff. I will do the testing now...

jan-petr reviewed (changes requested):
Deleted comment :)

jan-petr commented on Functions/xASL_test_Flavors.m:
```diff
@@ -1,26 +1,27 @@
-function xASL_test_Flavors(pathExploreASL, pathTest, bTest, x)
+function flavors = xASL_test_Flavors(testConfig, bTest, x, flavors)
```
`flavors` seems to be optional. But if I don't provide it, I get an error:

`Not enough input arguments.
Error in xASL_test_Flavors (line 74)
        flavors.loggingTable = xASL_test_Flavors_DCM2BIDS(testConfig, x, flavors.loggingTable);`

And it's not really clarified what should I input...

MichaelStritt replied:
Should be optional but really *recommended* right now 04b6f5a 👍

jan-petr commented on Functions/xASL_test_Flavors.m:
```diff
@@ -29,179 +30,84 @@ function xASL_test_Flavors(pathExploreASL, pathTest, bTest, x)
-defaultDataPar.x.settings.DELETETEMP = 1;
+    %% 1. Remove existing test data
+    if bTest(1)
+        xASL_test_Flavors_RemoveExistingTestData(testConfig);
```
If flavorList is not given in test-config, it all crashes. It would be nice - if flavor list is missing - to load by default all flavors.

MichaelStritt replied:
Ahhh okay, I understand. The thing is that this should work if you run the testConfig without flavorList from `xASL_test_FullPipelineTest`, because there it does this: 

```matlab
    % Check for flavor list
    if ~isfield(testConfig,'flavorList')
        testConfig.flavorList = xASL_adm_GetFileList(testConfig.pathFlavorDatabase, [], false, [], true);
    end
```

MichaelStritt replied:
But I can just copy paste that part to `xASL_test_Flavors.m` 👍

MichaelStritt replied:
Check this out 9315553 👍

MichaelStritt commented on Functions/xASL_bids_CompareStoreDifference.m:
```diff
@@ -0,0 +1,34 @@
+%
+% FORMAT: [identical,differences,dn] = xASL_bids_CompareStoreDifference(bPrintReport,differences,dn,allFiles,iFile)
+%
+% INPUT:
```
Ah, no worries, I can do that pretty quickly. I was still moving around the IO parameters a lot so I skipped this.

MichaelStritt replied:
Easy fix 04b6f5a 👍

MichaelStritt commented on Functions/xASL_bids_CompareStructures.m:
```diff
@@ -1,7 +1,7 @@
 %xASL_bids_CompareStructures Function that compares two BIDS folders with several subfolders and studies and prints the differences.
 %
-% FORMAT: [identical,results] = xASL_bids_CompareStructures(pathDatasetA,pathDatasetB,[bPrintReport,threshRmseNii]);
+% FORMAT: [identical,results,reportTable] = xASL_bids_CompareStructures(pathDatasetA,pathDatasetB,[bPrintReport,threshRmseNii]);
```
Only halfway 😆

MichaelStritt replied:
Easy fix 04b6f5a 👍

jan-petr commented on Functions/xASL_test_Flavors.m:
```diff
@@ -1,26 +1,27 @@
+%                        5. Check the the BIDS->Legacy import results
+%                        6. Run the ExploreASL on all datasets
+%                        7. Checks the ExploreASL processing results
+%                        (OPTIONAL, DEFAULT = [1 1 1 1 1 1 1])
```
The last 3 inputs are obviously not optional...
It doesn't work to run
xASL_test_Flavors(spm_jsonread('/home/janpetr/ExploreASL/ExploreASL/Testing/testConfig.json'),[],[],[])

We need to correctly check and initialized the defaults.

MichaelStritt replied:
Yep, so like above, I never had these problems since I mainly tested by running `xASL_test_FullPipelineTest`. I understand that this is not 100% correct though.
Does this 04b6f5a fix it already? I describe the last input as recommended and there is a fallback with a warning.

MichaelStritt replied:
x should be optional, there is a check for that.

MichaelStritt replied:
Same goes for bTest, I think you implemented that.

jan-petr commented on Functions/xASL_test_Flavors.m:
```diff
@@ -29,179 +30,84 @@ function xASL_test_Flavors(pathExploreASL, pathTest, bTest, x)
-end
+    %% 5. Run the comparison of data converted to the legacy format with the reference data
+    if bTest(5)
+        fprintf('\n================================= CHECK THE LEGACY CONVERSION ================================\n');
```
Complains about missing 
Warning: Missing file /ExploreASL/import_summary.csv  
We should add these CSV to exceptions - at least for checking BIDS2Legacy conversion

jan-petr replied:
This one as well|
Warning: Missing file /ExploreASL/bidsReportASL.json

jan-petr replied:
And the final logContent doesn't say anything about these - should we or should we not add the comparison differences to the final LogContent?

MichaelStritt replied:
Hey @jan-petr, yep saw those before. I was thinking that we could rather remove those from the flavors.comparisonTable instead. I thought that's not super high priority though, since we basically know about these things.
My wish would be to get rid of the logContent table and to mostly use flavors.comparisonTable and flavors.loggingTable instead. The overview there is a lot nicer and it works a in a more robust way. All of this log file scanning using regular expressions of XML and HTML text is super unstable.

MichaelStritt replied:
@jan-petr: check out this commit https://github.com/ExploreASL/ExploreASL/pull/835/commits/fd2fa5537078105a54d79dbf2ae5c4ee5355e142

HenkMutsaerts commented:
@MichaelStritt the system call on line 82 in `xASL_test_FullPipelineTest` can be improved, e.g.:

`[result1, result2] = system(...., '-echo');`

* `result1` & `result2` = 0 for correct system run and output in case of an error (can't remember the order :)
* `'echo'` ensures that the user sees on the screen what system does, rather than waiting a long time without seeing feedback

* compareStructures*: please write fully rather than using abbreviations, so I can also understand the code :) e.g., `dn`. Also variables like `identical` and `differences` are not self-explanatory?

* Note that we also had in our wishlist to:
  * Avoid copying `sourcedata`
  * Allow unzipping `sourcedata` on scantype level (new request from Paul Groot for efficient storage)
  * move some of the stuff that is in the test scripts here, to the actual ExploreASL import scripts. So the test script should run import scripts, not other test scripts
  * Make the import more modular (Michael & I should plan a meeting on this, perhaps good to discuss this physically next week)

jan-petr reviewed (changes requested):
Some more tiny issues, but other than that - it's smoothly running...

MichaelStritt commented on Functions/xASL_bids_CompareStructuresCheckContent.m:
```diff
@@ -0,0 +1,60 @@
+% FORMAT: [identical,differences] = xASL_bids_CompareStructuresCheckContent(filesDatasetA,filesDatasetB,pathDatasetA,pathDatasetB,identical,bPrintReport,detailedOutput,threshRmseNii)
+%
+% INPUT:
+%         ...
```
Easy fix 04b6f5a 👍

MichaelStritt commented:
@HenkMutsaerts & @jan-petr 

> @MichaelStritt the system call on line 82 in `xASL_test_FullPipelineTest` can be improved, e.g.:
> 
> `[result1, result2] = system(...., '-echo');`
> 
> * `result1` & `result2` = 0 for correct system run and output in case of an error (can't remember the order :)
> 
> * `'echo'` ensures that the user sees on the screen what system does, rather than waiting a long time without seeing feedback

I personally never used that option. I have my local flavor database that works nicely as is. Feel free to change this, but **IMHO** this is not part of this issue.

> * compareStructures*: please write fully rather than using abbreviations, so I can also understand the code :) e.g., `dn`. Also variables like `identical` and `differences` are not self-explanatory?

There are detailed explanations now **AND** those variables have been there before. This issue is **not** about optimizing the naming. It's about getting the flavor testing running again.

> Note that we also had in our wishlist to:
>       
> * Avoid copying `sourcedata`

Yes, that's what the new implementation does.

> * Allow unzipping `sourcedata` on scantype level (new request from Paul Groot for efficient storage)

I made a comment about that in the corresponding issue. I'm against that and/or don't want to implement that.

> * move some of the stuff that is in the test scripts here, to the actual ExploreASL import scripts. So the test script should run import scripts, not other test scripts

This is done step by step. Jan & I are getting rid of as much of the specific configurations as possible. Next steps here are to add `dataPar.jsons` where we need them instead of doing this in the testing script. As far as I know there are only like two or three flavors left that really require manual interaction **AND** all of this is **not** part of this issue/PR.

>  * Note my minor bugfix commit in [macOS test r1.8.0 import #842](https://github.com/ExploreASL/ExploreASL/issues/842).

Minor merge conflict incoming, but I think you'll be able to deal with this.

jan-petr reviewed (approved):
OK.
